### PR TITLE
Journeys

### PR DESCRIPTION
**Pull request by Megasonic988**

### CODE REVIEW

File: lib/udia/logs/journey.ex [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  alias Udia.Logs.Journey
+
+
+  schema "logs_journeys" do
+    field :description, :string
+    field :title, :string
+    belongs_to :explorer, Udia.Accounts.User
+    has_many :posts, Udia.Logs.Post
+
+    timestamps()
```
Comment (awwong1) [inline lib/udia/logs/journey.ex]: missing timestamps options for setting UTC. Default functionality is naive time, which we don't want.
`  @timestamps_opts [type: :utc_datetime, usec: true]`

File: lib/udia/logs/logs.ex [hunk 2]
```diff
@@ -204,4 +204,101 @@ defmodule Udia.Logs do
...
+
+      iex> delete_journey(journey)
+      {:ok, %Journey{}}
+
+      iex> delete_journey(journey)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def delete_journey(%Journey{} = journey) do
+    Repo.delete(journey)
```
Comment (awwong1) [inline lib/udia/logs/logs.ex]: Use `PaperTrail.delete`

File: lib/udia/web/controllers/journey_controller.ex [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  alias Udia.Logs
+  alias Udia.Logs.Journey
+
+  plug Guardian.Plug.EnsureAuthenticated, [handler: Udia.Web.SessionController] when action in [:create, :update, :delete]
+
+  action_fallback Udia.Web.FallbackController
+
+  def index(conn, _params) do
+    journeys = Logs.list_journeys()
```
Comment (awwong1) [inline lib/udia/web/controllers/journey_controller.ex]: This should be paginated

File: lib/udia/web/controllers/journey_controller.ex [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        |> put_status(:unprocessable_entity)
+        |> render(Udia.Web.ChangesetView, "error.json", changeset: changeset)
+
+    end
+  end
+
+  def show(conn, %{"id" => id}) do
+    journey = Logs.get_journey!(id)
+    journey = Udia.Repo.preload(journey, :explorer)
+    journey = Udia.Repo.preload(journey, :posts)
```
Comment (awwong1) [inline lib/udia/web/controllers/journey_controller.ex]: why does this not require `Udia.Repo.preload(journey, posts: :author)` like the create/index routes?

File: lib/udia/web/controllers/journey_controller.ex [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    if cur_user.id != journey.explorer_id do
+      conn
+      |> put_status(:forbidden)
+      |> render(Udia.Web.SessionView, "forbidden.json", error: "Invalid user")
+    else
+      case Logs.update_journey(cur_user, journey, journey_params) do
+        {:ok, journey_versioned} ->
+          journey = Map.get(journey_versioned, :model)
+          journey = Udia.Repo.preload(journey, :explorer)
+          journey = Udia.Repo.preload(journey, :posts)
```
Comment (awwong1) [inline lib/udia/web/controllers/journey_controller.ex]: again, what is the difference between `Udia.Repo.preload(journey, posts: :author)` and `Udia.Repo.preload(journey, :posts)`?

File: lib/udia/web/controllers/journey_controller.ex [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    render(conn, "index.json", journeys: journeys)
+  end
+
+  def create(conn, journey_params) do
+    cur_user = Guardian.Plug.current_resource(conn)
+    case Logs.create_journey(cur_user, journey_params) do
+      {:ok, journey_versioned} ->
+        journey = Map.get(journey_versioned, :model)
+        |> Udia.Repo.preload(:explorer)
+        |> Udia.Repo.preload(posts: :author)
```
Comment (awwong1) [inline lib/udia/web/controllers/journey_controller.ex]: I am hesitant to make preloading the posts default functionality for a journey. I envision a journey to have many posts- not sure if pagination like functionality is supported on children preload queries like you're doing here

File: lib/udia/web/views/journey_view.ex [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def render("journey.json", %{journey: journey}) do
+    %{id: journey.id,
+      title: journey.title,
+      description: journey.description,
+      explorer: render_one(journey.explorer, UserView, "user.json"),
+      inserted_at: journey.inserted_at,
+      updated_at: journey.updated_at,
+      posts: render_many(journey.posts, PostView, "post.json")}
```
Comment (awwong1) [inline lib/udia/web/views/journey_view.ex]: hesitant to do this, might be better to query for the posts using the journey.id instead (similar to how I'm doing the nested comments)

File: test/support/test_helpers.ex [hunk 8]
```diff
@@ -29,4 +29,24 @@ defmodule Udia.TestHelpers do
     {:ok, comment_versioned} = Udia.Logs.create_comment(user, changes)
     Map.get(comment_versioned, :model)
   end
+
+  def insert_post_with_journey(%Udia.Accounts.User{} = user, %Udia.Logs.Journey{} = journey, attrs \\%{}) do
```
Comment (awwong1) [inline test/support/test_helpers.ex]: just update the `insert_post` method to handle one more attr dude

File: test/udia/logs/logs_test.exs [hunk 9]
```diff
@@ -29,23 +30,35 @@ defmodule Udia.LogsTest do
       end
     end
 
-    test "create_post/1 with valid data creates a post" do
+    test "create_post/2 with valid data creates a post" do
```
Comment (awwong1) [inline test/udia/logs/logs_test.exs]: 👍  thanks for test arg number fix

File: test/udia/web/controllers/post_controller_test.exs [hunk 10]
```diff
@@ -1,8 +1,8 @@
 defmodule Udia.Web.PostControllerTest do
   use Udia.Web.ConnCase
 
-  @create_attrs %{content: "some content", title: "some title", type: "text"}
-  @update_attrs %{content: "some updated content", title: "some updated title"}
+  @create_attrs %{content: "some content", title: "some title", type: "text", journey: nil}
+  @update_attrs %{content: "some updated content", title: "some updated title", journey: nil}
```
Comment (awwong1) [inline test/udia/web/controllers/post_controller_test.exs]: Is this necessary? Seems strange to have to set the key even though the value is `nil`

**Comment by awwong1** [comment]
Add test for `journey_controller.show/2` to satisfy coverage. 
https://coveralls.io/builds/12489759/source?filename=lib%2Fudia%2Fweb%2Fcontrollers%2Fjourney_controller.ex
Slight concern about loading the posts, as mentioned in the inline comments

File: lib/udia/web/controllers/journey_controller.ex (hunk 5 above)
Comment (Megasonic988) [inline lib/udia/web/controllers/journey_controller.ex]: the difference is the author nested in posts is preloaded as well in this case

File: lib/udia/logs/journey.ex (hunk 1 above)
Comment (Megasonic988) [inline lib/udia/logs/journey.ex]: fixed

File: lib/udia/logs/logs.ex (hunk 2 above)
Comment (Megasonic988) [inline lib/udia/logs/logs.ex]: fixed

File: lib/udia/web/controllers/journey_controller.ex (hunk 4 above)
Comment (Megasonic988) [inline lib/udia/web/controllers/journey_controller.ex]: fixed

File: lib/udia/web/controllers/journey_controller.ex (hunk 5 above)
Comment (Megasonic988) [inline lib/udia/web/controllers/journey_controller.ex]: fixed

File: lib/udia/web/views/journey_view.ex (hunk 7 above)
Comment (Megasonic988) [inline lib/udia/web/views/journey_view.ex]: fixed

File: test/udia/web/controllers/post_controller_test.exs (hunk 10 above)
Comment (Megasonic988) [inline test/udia/web/controllers/post_controller_test.exs]: fixed

File: test/support/test_helpers.ex (hunk 8 above)
Comment (Megasonic988) [inline test/support/test_helpers.ex]: fixed

File: test/udia/logs/logs_test.exs (hunk 9 above)
Comment (Megasonic988) [inline test/udia/logs/logs_test.exs]: fixed

File: lib/udia/web/controllers/journey_controller.ex (hunk 6 above)
Comment (Megasonic988) [inline lib/udia/web/controllers/journey_controller.ex]: fixed

File: lib/udia/web/controllers/journey_controller.ex (hunk 3 above)
Comment (Megasonic988) [inline lib/udia/web/controllers/journey_controller.ex]: fixed

**Pull request merged**